src/scalar.rs

```diff
@@ -894,449 +540,807 @@ impl<'a> From<&'a Scalar> for [u8; Scalar::SIZE] {
-                8, 216, 57, 51, 72, 125, 157, 41, 83, 167, 237, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-            ])
-        );
+#[test]
+fn test_subtraction() {
+    let mut tmp = LARGEST;
+    tmp -= &LARGEST;
+
+    assert_eq!(tmp, Scalar::zero());
+
+    let mut tmp = Scalar::zero();
+    tmp -= &LARGEST;
+
+    let mut tmp2 = MODULUS;
+    tmp2 -= &LARGEST;
+
+    assert_eq!(tmp, tmp2);
+}
+
+#[test]
+fn test_multiplication() {
+    let mut cur = LARGEST;
+
+    for _ in 0..100 {
+        let mut tmp = cur;
+        tmp *= &cur;
+
+        let mut tmp2 = Scalar::zero();
+        for b in cur
+            .to_bytes()
+            .iter()
+            .rev()
+            .flat_map(|byte| (0..8).rev().map(move |i| ((byte >> i) & 1u8) == 1u8))
+        {
+            let tmp3 = tmp2;
+            tmp2.add_assign(&tmp3);
+
+            if b {
+                tmp2.add_assign(&cur);
+            }
+        }
+
+        assert_eq!(tmp, tmp2);
+
+        cur.add_assign(&LARGEST);
     }
+}
 
-    #[test]
-    fn test_from_bytes_wide_maximum() {
-        assert_eq!(
-            Scalar([
-                0xc62c1805439b73b1,
-                0xc2b9551e8ced218e,
-                0xda44ec81daf9a422,
-                0x5605aa601c162e79
-            ]),
-            Scalar::from_bytes_wide(&[0xff; 64])
-        );
+#[test]
+fn test_squaring() {
+    let mut cur = LARGEST;
+
+    for _ in 0..100 {
+        let mut tmp = cur;
+        tmp = tmp.square();
+
+        let mut tmp2 = Scalar::zero();
+        for b in cur
+            .to_bytes()
+            .iter()
+            .rev()
+            .flat_map(|byte| (0..8).rev().map(move |i| ((byte >> i) & 1u8) == 1u8))
+        {
+            let tmp3 = tmp2;
+            tmp2.add_assign(&tmp3);
+
+            if b {
+                tmp2.add_assign(&cur);
+            }
+        }
+
+        assert_eq!(tmp, tmp2);
+
+        cur.add_assign(&LARGEST);
     }
+}
 
-    #[test]
-    fn test_zero() {
-        assert_eq!(Scalar::zero(), -&Scalar::zero());
-        assert_eq!(Scalar::zero(), Scalar::zero() + Scalar::zero());
-        assert_eq!(Scalar::zero(), Scalar::zero() - Scalar::zero());
-        assert_eq!(Scalar::zero(), Scalar::zero() * Scalar::zero());
+#[test]
+fn test_inversion() {
+    assert_eq!(Scalar::zero().invert(), None);
+    assert_eq!(Scalar::one().invert().unwrap(), Scalar::one());
+    assert_eq!((-&Scalar::one()).invert().unwrap(), -&Scalar::one());
+
+    let mut tmp = R2;
+
+    for _ in 0..100 {
+        let mut tmp2 = tmp.invert().unwrap();
+        tmp2.mul_assign(&tmp);
+
+        assert_eq!(tmp2, Scalar::one());
+
+        tmp.add_assign(&R2);
     }
+}
 
-    #[cfg(test)]
-    const LARGEST: Scalar = Scalar([
-        0xffffffff00000000,
+#[test]
+fn test_invert_is_pow() {
+    let q_minus_2 = [
+        0xfffffffeffffffff,
```

Comment:
Why are these numbers not separated by `_`?

Reply:
Good catch, they should! Will fix